src/main/java/io/odpf/firehose/sink/objectstorage/writer/remote/ObjectStorageChecker.java

```diff
@@ -1,52 +1,131 @@
     private final Set<ObjectStorageWriterWorkerFuture> remoteUploadFutures;
     private final ExecutorService remoteUploadScheduler;
     private final ObjectStorage objectStorage;
+    private final Clock clock;
+    private final Instrumentation instrumentation;
 
     @Override
     public void run() {
-        List<String> tobeFlushed = new ArrayList<>();
+        List<FileMeta> tobeFlushed = new ArrayList<>();
         toBeFlushedToRemotePaths.drainTo(tobeFlushed);
         remoteUploadFutures.addAll(tobeFlushed.stream()
-                .map(path -> new ObjectStorageWriterWorkerFuture(
-                        remoteUploadScheduler.submit(() -> objectStorage.store(path)), path)
+                .map(fileMeta -> {
+                            Runnable uploadTask = createUploadTask(fileMeta);
+                            Future<?> future = remoteUploadScheduler.submit(uploadTask);
+                            Instant startTime = clock.now();
+                            return new ObjectStorageWriterWorkerFuture(future, fileMeta, startTime);
+                        }
                 ).collect(Collectors.toList()));
 
-        Set<String> flushedPath = remoteUploadFutures.stream().map(future -> {
-            if (!future.getFuture().isDone()) {
+        Set<String> flushedPath = remoteUploadFutures.stream().map(uploadJob -> {
+            if (!uploadJob.getFuture().isDone()) {
                 return "";
             } else {
                 try {
-                    future.getFuture().get();
-                    LOGGER.info("Flushed to Object storage " + future.getPath());
-                } catch (InterruptedException e) {
-                    throw new ObjectStorageFailedException(e);
-                } catch (ExecutionException e) {
-                    throw new ObjectStorageFailedException(e.getCause());
+                    uploadJob.getFuture().get();
+                    FileMeta fileMeta = uploadJob.getFileMeta();
+                    LOGGER.info("Flushed to Object storage " + fileMeta.getFullPath());
+                    captureFileUploadSuccessMetric(uploadJob, fileMeta);
+                } catch (InterruptedException | ExecutionException e) {
+                    captureUploadFailedMetric(uploadJob, e);
+                    throw new RuntimeException(e);
                 }
-                return future.getPath();
+
+                return uploadJob.getFileMeta().getFullPath();
             }
         }).filter(x -> !x.isEmpty()).collect(Collectors.toSet());
-        remoteUploadFutures.removeIf(x -> flushedPath.contains(x.getPath()));
+        remoteUploadFutures.removeIf(x -> flushedPath.contains(x.getFileMeta().getFullPath()));
         flushedToRemotePaths.addAll(flushedPath);
     }
+
+    private Runnable createUploadTask(FileMeta fileMeta) {
+        return () -> {
+            try {
+                objectStorage.store(fileMeta.getFullPath());
+            } catch (ObjectStorageException e) {
```

Comment:
ahh. we don't need this! ObjectStorageException is already a RuntimeException. 
no need to change this line.